File: backend/services/weather-service.ts
import axios from "axios";

interface WeatherData {
  clouds: { all: number };
  main: { humidity: number; temp: number };
  weather: Array<{ main: string; description: string }>;
  wind: { speed: number };
  visibility: number;
}

interface VisibilityPrediction {
  score: number;
  level: "excellent" | "good" | "fair" | "poor";
  message: string;
  details: {
    cloudCover: number;
    humidity: number;
    conditions: string[];
  };
}

export class WeatherService {
  private apiKey: string;
  private baseUrl = "https://api.openweathermap.org/data/2.5";

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  async getWeatherData(lat: number, lng: number): Promise<WeatherData> {
    try {
      const response = await axios.get(`${this.baseUrl}/weather`, {
        params: {
          lat,
          lon: lng,
          appid: this.apiKey,
          units: "metric",
          lang: "ja",
        },
      });
      return response.data;
    } catch (error) {
      throw new Error(`天気データの取得に失敗しました: ${error}`);
    }
  }

  calculateVisibility(weather: WeatherData): VisibilityPrediction {
    let score = 0;
    const conditions: string[] = [];

    // 雲量チェック（最重要）
    const cloudCover = weather.clouds.all;
    if (cloudCover >= 30 && cloudCover <= 70) {
      score += 40;
      conditions.push("適度な雲量");
    } else if (cloudCover < 30) {
      score += 10;
      conditions.push("雲が少なめ");
    } else if (cloudCover > 80) {
      score -= 20;
      conditions.push("雲が多い");
    }

    // 湿度チェック
    const humidity = weather.main.humidity;
    if (humidity >= 40 && humidity <= 70) {
      score += 25;
      conditions.push("適度な湿度");
    } else if (humidity > 70) {
      score += 15;
      conditions.push("湿度高め");
    } else {
      score += 5;
      conditions.push("乾燥気味");
    }

    // 天気状況チェック
    const mainWeather = weather.weather[0].main;
    if (mainWeather === "Clear") {
      score += 10;
      conditions.push("晴れ");
    } else if (mainWeather === "Clouds") {
      score += 20;
      conditions.push("曇り");
    } else if (mainWeather === "Rain") {
      score -= 30;
      conditions.push("雨");
    }

    // 風速チェック（空気の澄み具合）
    const windSpeed = weather.wind.speed;
    if (windSpeed >= 2 && windSpeed <= 5) {
      score += 10;
      conditions.push("適度な風");
    }

    // 視程チェック
    if (weather.visibility >= 10000) {
      score += 15;
      conditions.push("視程良好");
    }

    return this.getVisibilityLevel(score, cloudCover, humidity, conditions);
  }

  private getVisibilityLevel(
    score: number,
    cloudCover: number,
    humidity: number,
    conditions: string[]
  ): VisibilityPrediction {
    let level: VisibilityPrediction["level"];
    let message: string;

    if (score >= 80) {
      level = "excellent";
      message = "✨ 絶好のマジックアワー日和です！";
    } else if (score >= 60) {
      level = "good";
      message = "👌 美しい空が期待できそうです";
    } else if (score >= 40) {
      level = "fair";
      message = "🤔 見える可能性はありますが...";
    } else {
      level = "poor";
      message = "😔 今日は厳しそうです";
    }

    return {
      score,
      level,
      message,
      details: {
        cloudCover,
        humidity,
        conditions,
      },
    };
  }
}

export type { WeatherData, VisibilityPrediction };

File: backend-python/services/weather_service.py
import os
import requests
from typing import Dict, Any, Optional
from datetime import datetime
import asyncio
from concurrent.futures import ThreadPoolExecutor


class WeatherService:
    def __init__(self):
        self.api_key = os.getenv("OPENWEATHER_API_KEY")
        self.base_url = "https://api.openweathermap.org/data/2.5"
        self.executor = ThreadPoolExecutor(max_workers=3)
    
    async def get_weather(self, latitude: float, longitude: float) -> Dict[str, Any]:
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            self.executor,
            self._fetch_weather,
            latitude,
            longitude
        )
    
    def _fetch_weather(self, latitude: float, longitude: float) -> Dict[str, Any]:
        if not self.api_key:
            raise ValueError("OpenWeather API key not found in environment variables")
        
        url = f"{self.base_url}/weather"
        params = {
            "lat": latitude,
            "lon": longitude,
            "appid": self.api_key,
            "units": "metric"
        }
        
        try:
            response = requests.get(url, params=params, timeout=10)
            response.raise_for_status()
            data = response.json()
            
            return self._format_weather_data(data)
        except requests.RequestException as e:
            raise Exception(f"Weather API error: {str(e)}")
    
    def _format_weather_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
        return {
            "temperature": data["main"]["temp"],
            "feels_like": data["main"]["feels_like"],
            "humidity": data["main"]["humidity"],
            "pressure": data["main"]["pressure"],
            "wind_speed": data["wind"]["speed"],
            "wind_direction": data["wind"].get("deg", 0),
            "clouds": data["clouds"]["all"],
            "visibility": data.get("visibility", 10000),
            "weather": {
                "main": data["weather"][0]["main"],
                "description": data["weather"][0]["description"],
                "icon": data["weather"][0]["icon"]
            },
            "location": data["name"],
            "country": data["sys"]["country"],
            "sunrise": datetime.fromtimestamp(data["sys"]["sunrise"]).isoformat(),
            "sunset": datetime.fromtimestamp(data["sys"]["sunset"]).isoformat(),
            "timestamp": datetime.fromtimestamp(data["dt"]).isoformat()
        }
    
    async def get_forecast(self, latitude: float, longitude: float, days: int = 5) -> Dict[str, Any]:
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            self.executor,
            self._fetch_forecast,
            latitude,
            longitude,
            days
        )
    
    def _fetch_forecast(self, latitude: float, longitude: float, days: int) -> Dict[str, Any]:
        if not self.api_key:
            raise ValueError("OpenWeather API key not found in environment variables")
        
        url = f"{self.base_url}/forecast"
        params = {
            "lat": latitude,
            "lon": longitude,
            "appid": self.api_key,
            "units": "metric",
            "cnt": days * 8
        }
        
        try:
            response = requests.get(url, params=params, timeout=10)
            response.raise_for_status()
            data = response.json()
            
            return self._format_forecast_data(data)
        except requests.RequestException as e:
            raise Exception(f"Forecast API error: {str(e)}")
    
    def _format_forecast_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
        forecasts = []
        for item in data["list"]:
            forecasts.append({
                "datetime": datetime.fromtimestamp(item["dt"]).isoformat(),
                "temperature": item["main"]["temp"],
                "feels_like": item["main"]["feels_like"],
                "humidity": item["main"]["humidity"],
                "clouds": item["clouds"]["all"],
                "wind_speed": item["wind"]["speed"],
                "weather": {
                    "main": item["weather"][0]["main"],
                    "description": item["weather"][0]["description"],
                    "icon": item["weather"][0]["icon"]
                },
                "precipitation": item.get("rain", {}).get("3h", 0) + item.get("snow", {}).get("3h", 0)
            })
        
        return {
            "location": data["city"]["name"],
            "country": data["city"]["country"],
            "forecasts": forecasts
        }